mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -13,25 +17,104 @@ impl S3CrtClient {
         key: &str,
         params: &PutObjectParams,
     ) -> ObjectClientResult<S3PutObjectRequest, PutObjectError, S3RequestError> {
+        Ok(S3PutObjectRequest::new(
+            self.inner.clone(),
+            bucket.to_owned(),
+            key.to_owned(),
+            params.to_owned(),
+        ))
+    }
+}
+
+type S3PutRequest = (S3HttpRequest<Vec<u8>, PutObjectError>, AsyncStreamWriter);
+
+#[derive(Debug)]
+pub struct S3PutObjectRequest {
+    client: Arc<S3CrtClientInner>,
+    bucket: String,
+    key: String,
+    params: PutObjectParams,
+    inner: Option<S3PutRequest>,
+}
+
+impl S3PutObjectRequest {
+    fn new(client: Arc<S3CrtClientInner>, bucket: String, key: String, params: PutObjectParams) -> Self {
+        Self {
+            client,
+            bucket,
+            key,
+            params,
+            inner: Default::default(),
+        }
+    }
+
+    fn make_request(&self, content_length: Option<usize>) -> Result<S3PutRequest, S3RequestError> {
```

Comment:
I wonder if things would be clearer with two separate methods, one for zero-size objects and one for everything else. The zero-size version maybe doesn't need to create/set the body stream at all, which might make the `None` case in `complete` clearer.

I actually wonder if we should even go through the `MetaRequestType::PutObject` machinery at all for zero-byte objects, instead preferring to just to a basic HTTP PutObject request. But that's a premature optimization we can revisit later.

Reply:
I did experiment with that, but the CRT complains if we do not set any body, even though Content-Length==0: 
```
Could not create auto-ranged-put meta request. Body must be set via filepath, async stream, or body stream. 
```
There are certainly ways around that, but I went for the minimal change to work around the current CRT limitation, assuming it will be removed soon. We should definitely review if that doesn't happen and/or we want to extend the client. 
For now, I will try and make it clearer in comments.